File: solutions/fsharp/allergies/1/Allergies.fs
module Allergies

type Allergen = 
| Eggs = 1
| Peanuts = 2
| Shellfish = 4
| Strawberries = 8
| Tomatoes = 16
| Chocolate = 32
| Pollen = 64
| Cats = 128

let allergicTo (codedAllergies: int) (allergen: Allergen): bool =
    0 < ((int allergen) &&& codedAllergies)

let allAllergens: List<Allergen> = 
    System.Enum.GetValues(typeof<Allergen>) |> Seq.cast<Allergen> |> List.ofSeq

let list (codedAllergies: int): List<Allergen> = 
     allAllergens |> List.filter (allergicTo codedAllergies)

File: solutions/nim/allergies/1/allergies.nim
import std/strformat
  
type
  Allergen* = enum
    Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  (score and cast[int]({allergen})) > 0

proc allergies*(score: int): set[Allergen] =
  cast[set[Allergen]](score)
